rclrs/src/node/client.rs

```diff
@@ -0,0 +1,275 @@
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
+            )
+        }
+        .ok()?;
+        let (tx, rx) = oneshot::channel::<T::Response>();
+        self.futures.lock().insert(sequence_number, tx);
+        // It is safe to call unwrap() here since the `Canceled` error will only happen when the
+        // `Sender` is dropped
+        // https://docs.rs/futures/latest/futures/channel/oneshot/struct.Canceled.html
+        Ok(rx.await.unwrap())
+    }
+
+    /// Fetches a new response.
+    ///
+    /// When there is no new message, this will return a
+    /// [`ClientTakeFailed`][1].
+    ///
+    /// [1]: crate::RclrsError
+    //
+    // ```text
+    // +----------------------+
+    // | rclrs::take_response |
+    // +----------+-----------+
+    //            |
+    //            |
+    // +----------v-----------+
+    // |   rcl_take_response  |
+    // +----------+-----------+
+    //            |
+    //            |
+    // +----------v----------+
+    // |      rmw_take       |
+    // +---------------------+
+    // ```
+    pub fn take_response(&self) -> Result<(T::Response, rmw_request_id_t), RclrsError> {
+        let mut request_id_out = rmw_request_id_t {
+            writer_guid: [0; 16],
+            sequence_number: 0,
+        };
+        type RmwMsg<T> =
+            <<T as rosidl_runtime_rs::Service>::Response as rosidl_runtime_rs::Message>::RmwMsg;
+        let mut response_out = RmwMsg::<T>::default();
+        let handle = &*self.handle.lock();
+        unsafe {
+            // SAFETY: The three pointers are valid/initialized
+            rcl_take_response(
+                handle,
+                &mut request_id_out,
+                &mut response_out as *mut RmwMsg<T> as *mut _,
+            )
+        }
+        .ok()?;
+        Ok((T::Response::from_rmw_message(response_out), request_id_out))
+    }
+}
+
+impl<T> ClientBase for Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    fn handle(&self) -> &ClientHandle {
+        &self.handle
+    }
+
+    fn execute(&self) -> Result<(), RclrsError> {
+        let (res, req_id) = match self.take_response() {
+            Ok((res, req_id)) => (res, req_id),
+            Err(RclrsError::RclError {
+                code: RclReturnCode::ClientTakeFailed,
+                ..
+            }) => {
+                // Spurious wakeup – this may happen even when a waitset indicated that this
+                // client was ready, so it shouldn't be an error.
+                return Ok(());
+            }
+            Err(e) => return Err(e),
+        };
+        let requests = &mut *self.requests.lock();
+        let futures = &mut *self.futures.lock();
+        if let Some(callback) = requests.remove(&req_id.sequence_number) {
+            callback(res);
+        } else if let Some(future) = futures.remove(&req_id.sequence_number) {
+            future.send(res).unwrap_or(());
```

Comment:
I believe the canonical way to ignore an error is:

```suggestion
            let _ = future.send(res);
```

but no need to change it if you like this way more.

Reply:
Done in bf61831